File: aws/CloudFormation/Sequence/sequence.go
package main

import (
	"context"
	"github.com/aws/aws-lambda-go/cfn"
	"github.com/aws/aws-lambda-go/lambda"
	"github.com/aws/aws-sdk-go-v2/aws/external"
	awsssm "github.com/aws/aws-sdk-go-v2/service/ssm"
	"github.com/codesmith-gmbh/forge/aws/common"
	"github.com/mitchellh/mapstructure"
	"github.com/pkg/errors"
	"strings"
)

type proc struct {
	ssm *awsssm.SSM
}

func main() {
	cfg, err := external.LoadDefaultAWSConfig()
	if err != nil {
		panic(err)
	}
	p := &proc{ssm: awsssm.New(cfg)}
	lambda.Start(cfn.LambdaWrap(p.processEvent))
}

type Properties struct {
	SequenceName, Expression string
}

func properties(input map[string]interface{}) (Properties, error) {
	var properties Properties
	if err := mapstructure.Decode(input, &properties); err != nil {
		return properties, err
	}
	if !strings.HasPrefix(properties.SequenceName, "/") {
		return properties, errors.Errorf("name %s must start with an /", properties.SequenceName)
	}
	if properties.Expression == "" {
		properties.Expression = "x"
	}
	if _, err := common.Eval(properties.Expression, 1); err != nil {
		return properties, err
	}
	return properties, nil
}

func (p *proc) processEvent(ctx context.Context, event cfn.Event) (string, map[string]interface{}, error) {
	properties, err := properties(event.ResourceProperties)
	if err != nil {
		return "", nil, err
	}
	switch event.RequestType {
	case cfn.RequestDelete:
		if common.IsForgeSsmParameterARN(event.PhysicalResourceID) {
			_, err := p.ssm.DeleteParameterRequest(&awsssm.DeleteParameterInput{
				Name: &event.PhysicalResourceID,
			}).Send(ctx)
			if err != nil {
				return event.PhysicalResourceID, nil, errors.Wrapf(err, "could not delete the Sequence %s", properties.SequenceName)
			}
		}
		return event.PhysicalResourceID, nil, nil
	case cfn.RequestCreate:
		return p.putSequence(ctx, properties)
	case cfn.RequestUpdate:
		return p.putSequence(ctx, properties)
	default:
		return "", nil, errors.Errorf("unknown request type %s", event.RequestType)
	}
}

func (p *proc) putSequence(ctx context.Context, properties Properties) (string, map[string]interface{}, error) {
	overwrite := true
	parameterName := common.SequenceParameterName(properties.SequenceName)
	_, err := p.ssm.PutParameterRequest(&awsssm.PutParameterInput{
		Name:      &parameterName,
		Type:      awsssm.ParameterTypeString,
		Value:     &properties.Expression,
		Overwrite: &overwrite,
	}).Send(ctx)
	if err != nil {
		return "", nil, errors.Wrapf(err, "could not put the parameter %s", parameterName)
	}
	return properties.SequenceName, nil, nil
}

File: aws/CloudFormation/Sequence/sequence.py
import logging

import boto3
from box import Box
from crhelper import CfnResource
from schema import And, Optional, Schema

from aws.common import naming
from aws.common.calc import calculator, SSM_PARAMETER_DESCRIPTION
from aws.common.cfn import resource_properties
from aws.common.schema import not_empty
from aws.common.ssm import silent_delete_parameter_from_event, put_string_parameter

helper = CfnResource()
logger = logging.getLogger(__name__)
logger.setLevel(logging.INFO)

ssm = boto3.client('ssm')
expression_parser = calculator(1)

properties_schema = Schema({
    'SequenceName': And(str, not_empty, error='not empty string for SequenceName'),
    Optional('Expression'): And(str, not_empty, error='not empty string for Expression')
})


def validate_properties(properties):
    p = properties_schema.validate(properties)
    if p.expression is None:
        p.expression = 'x'
    expression_parser.parse(p.expression)
    return Box(p, camel_killer_box=True)


@helper.create
@helper.update
def create(event, _):
    properties = validate_properties(resource_properties(event))
    return put_sequence_parameter(properties)


def put_sequence_parameter(properties):
    parameter_name = naming.sequence_parameter_name(properties.parameter_name)
    return put_string_parameter(ssm, parameter_name,
                                value=properties.expression,
                                description=SSM_PARAMETER_DESCRIPTION)


@helper.delete
def delete(event, _):
    return silent_delete_parameter_from_event(ssm, event)


def handler(event, context):
    logger.info('event: %s', event)
    helper(event, context)
